service/storage_service.cc

```diff
@@ -773,20 +797,52 @@ void storage_service::join_token_ring(int delay) {
         MigrationManager.announceNewKeyspace(TraceKeyspace.definition(), 0, false);
 #endif
 
-    // At this point our local tokens are chosen (_bootstrap_tokens) and will not be changed unless we bootstrap again.
-    db::system_keyspace::set_bootstrap_state(db::system_keyspace::bootstrap_state::COMPLETED).get();
-
     slogger.debug("Setting tokens to {}", _bootstrap_tokens);
     // This node must know about its chosen tokens before other nodes do
     // since they may start sending writes to this node after it gossips status = NORMAL.
     // Therefore, in case we haven't updated _token_metadata with our tokens yet, do it now.
     _token_metadata.update_normal_tokens(_bootstrap_tokens, get_broadcast_address());
 
+    if (!_streams_ts && db().local().get_config().check_experimental(db::experimental_features_t::CDC)) {
```

Comment:
If you take out the comments, there is very little code.
The "meat" of CDC code is in CDC-specific functions like `make_new_cdc_generation`. I will move their implementations to a different file in the cdc/ directory.

But this particular piece of code should be inlined here in `join_token_ring`, because to understand what it does and why it's here, you need to understand `join_token_ring` (and consequently, `prepare_to_join` and `init_server`... you probably need to understand 2/3 of storage_service.cc to understand how nodes start). There's a bunch of "invisible" relationships and invariants that must hold while this particular piece of code runs.

The comments are there to explain these relationships (I wish there were many more such comments in other parts of `join_token_ring` and `prepare_to_join`).

If I were to move this code to a different file, I would have to write more explanations ("this procedure must be called exactly from that place or it won't work"). And then to understand it you would have to open both storage_service.cc and the other file in a vertical split and read them consecutively.